edx_arch_experiments/codejail_service/views.py

```diff
@@ -0,0 +1,153 @@
+        # Some of these are configured as union types because
+        # edx-platform appears to currently default to None for some
+        # of them (rather than omitting the keys.)
+        'python_path': {
+            'anyOf': [
+                {
+                    'type': 'array',
+                    'items': {'type': 'string'},
+                },
+                {'type': 'null'},
+            ],
+        },
+        'limit_overrides_context': {
+            'anyOf': [
+                {'type': 'string'},
+                {'type': 'null'},
+            ],
+        },
+        'slug': {
+            'anyOf': [
+                {'type': 'string'},
+                {'type': 'null'},
+            ],
+        },
+        'unsafely': {'type': 'boolean'},
+    },
+    'required': ['code'],
+}
+
+# A note on the authorization model used here:
+#
+# We really just need one service account to be able to call this, and
+# then also allow is_staff to call it for convenience and debugging
+# purposes.
+#
+# To do this "right", I'd probably have to create an empty abstract
+# model, create a permission on it, create a group, add the permission
+# to the group, and add the service account to the group. Then I could
+# check the calling user's has_perm. If I wanted to use bridgekeeper
+# (as we're trying to do more of) I might be able to give bridgekeeper
+# a `@blanket_rule` that checks membership in the group, then use
+# bridgekeeper here instead of checking permissions directly, but it's
+# possible this wouldn't work because bridgekeeper might require there
+# to be a model instance to pass in (and there wouldn't be one, since
+# it's just an abstract model.)
+#
+# But... given that the service account will be is_staff, and I'll be
+# opening this up to is_staff alongside the intended service account,
+# and this is already a hacky intermediate solution... we can just use
+# the `IsAdminUser` permission class and be done with it.
+
+
+@api_view(['POST'])
+@parser_classes([FormParser, MultiPartParser])
+@permission_classes([IsAdminUser])
+def code_exec_view_v0(request):
+    """
+    Executes code in a codejail sandbox for a remote caller.
+
+    This implements the API used by edxapp's xmodule.capa.safe_exec.remote_exec.
+    It accepts a POST of a form containing a `payload` value and zero or more
+    extra files.
+
+    The payload is JSON and contains the parameters to be sent to codejail's
+    safe_exec (aside from `extra_files`). See payload_schema for type information.
+
+    This API does not permit `unsafely=true`.
+    """
+    if not CODEJAIL_SERVICE_ENABLED.is_enabled():
+        return Response("Codejail service not enabled", status=500)
+
+    params_json = request.data['payload']
+    params = json.loads(params_json)
+    jsonschema.validate(params, payload_schema)
+
+    complete_code = params['code']  # includes standard prolog
+    input_globals_dict = params.get('globals_dict') or {}
```

Comment:
Do we have to use `or` here because `params.get('globals_dict', {})` will return `None` if `None` is the actual value of the key?

Reply:
Oh, hmm. That was the _idea_, but looking again at [safe_exec.py in edx-platform](https://github.com/openedx/edx-platform/blob/4a1458b9d69462429bc64d2b077730f77bd8cd3c/xmodule/capa/safe_exec/safe_exec.py#L84) (and the calling code in ../capa_problem.py) I see that I had misread; globals_dict is _not_ defaulted to None, and will always be a dictionary (because it's functionally a return argument, and is updated in-place).

I've also set the schema to disallow an explicit None for globals_dict, so this line can already be safely changed to `params['globals_dict']`. I'll make that change and I'll add globals_dict to the required-fields list in the schema.